src/Grpc.AspNetCore.Server/GrpcServiceExtensions.cs

```diff
@@ -63,11 +65,15 @@ public static IGrpcServerBuilder AddGrpc(this IServiceCollection services)
             services.AddRouting();
             services.AddOptions();
             services.TryAddSingleton<GrpcMarkerService>();
-            services.TryAddSingleton<ServiceMethodsRegistry>();
             services.TryAddSingleton(typeof(ServerCallHandlerFactory<>));
             services.TryAddScoped(typeof(IGrpcServiceActivator<>), typeof(DefaultGrpcServiceActivator<>));
             services.TryAddScoped(typeof(IGrpcInterceptorActivator<>), typeof(DefaultGrpcInterceptorActivator<>));
             services.TryAddEnumerable(ServiceDescriptor.Singleton<IConfigureOptions<GrpcServiceOptions>, GrpcServiceOptionsSetup>());
+            
+            // Model
+            services.TryAddSingleton<ServiceMethodsRegistry>();
+            services.TryAddSingleton(typeof(ServiceRouteBuilder<>));
+            services.TryAddEnumerable(ServiceDescriptor.Singleton(typeof(IServiceMethodProvider<>), typeof(BinderServiceMethodProvider<>)));
```

Comment:
> If there are multiple IServiceMethodProviders added, how will we ensure the same method isn't added multiple times?

We don't. This is designed to be a very simple way to inject logic for resolving methods from a service type. Order doesn't matter because everyone is only adding.